From: Sven Eckelmann <sven@narfation.org>
Date: Sat, 24 Jul 2021 17:47:13 +0200
Subject: asm-generic: avoid sparse {get,put}_unaligned warning

Sparse will try to check casting of simple integer types which are marked
as __bitwise. This for example "disallows" simple casting of __be{16,32,64}
or __le{16,32,64} to other types. This is also true for pointers to
variables with this type.

But the new generic {get,put}_unaligned is doing that by (reinterpret)
casting the original pointer to a new (anonymous) struct pointer. This will
then create warnings like:

  net/batman-adv/distributed-arp-table.c:1461:19: warning: cast from restricted __be32 *
  net/batman-adv/distributed-arp-table.c:1510:23: warning: cast from restricted __be32 [usertype] *[assigned] magic
  net/batman-adv/distributed-arp-table.c:1588:24: warning: cast from restricted __be32 [usertype] *[assigned] yiaddr

The special attribute force must be used in such statements when the cast
is known to be safe to avoid these warnings.

Fixes: 803f4e1eab7a ("asm-generic: simplify asm/unaligned.h")
Signed-off-by: Sven Eckelmann <sven@narfation.org>

diff --git a/include/vdso/unaligned.h b/include/vdso/unaligned.h
index ff0c06b6513eff61046f9fcf40111c9a8baf176b..643a6dbfbddf0dc3377b0dbbcfd9461632920d2f 100644
--- a/include/vdso/unaligned.h
+++ b/include/vdso/unaligned.h
@@ -3,12 +3,13 @@
 #define __VDSO_UNALIGNED_H
 
 #define __get_unaligned_t(type, ptr) ({							\
-	const struct { type x; } __packed * __get_pptr = (typeof(__get_pptr))(ptr);	\
+	const struct { type x; } __packed * __get_pptr;					\
+	__get_pptr = (__force typeof(__get_pptr))(ptr);					\
 	__get_pptr->x;									\
 })
 
 #define __put_unaligned_t(type, val, ptr) do {						\
-	struct { type x; } __packed * __put_pptr = (typeof(__put_pptr))(ptr);		\
+	struct { type x; } __packed * __put_pptr = (__force typeof(__put_pptr))(ptr);	\
 	__put_pptr->x = (val);								\
 } while (0)
 
